### ISSUE TITLE
Resolution is incorrectly handled in 5.x

### ISSUE THREAD

**Issue by butlerpd** [SasView/sasview#1547]
Pursuant to observations form an observant user it was discovered that the resolution smearing is being done completely wrong in 5.x, using a single value of dq/q to apply to all points. Specifically @gonzalezma found that the same function (`onPinholeSmear`) in `SmearingWidget.py`,  is called regardless of whether the Custom Pinhole Smear or Use dQ Data option are chosen.  Using dQ Data, the % is calculated from the [dQ/Q]max value obtained from the data, so that the smearing seems to work unless one looks closely, but the same % is erroneously applied to all the Q points. Moreover, when swapping between choosing Using dQ Data and Custom Pinhole, the values of of [dQ/Q]min and [dQ/Q]max seemed to be reversed.

**Comment by gonzalezma** [comment]
The branch ESS_GUI_1547_smearing contains a possible workaround, but it should be checked for 2D as some things were not very clear in the code and the interface, already in SasView 4.
The 'Custom Pinhole Smear' interface for 2D indicates <dQp> and <dQs>, but there is a single input box. How should this be handled? In this commit I simply assumed that there is a single input value corresponding to 100*dq/q (as in the 1D case) and from the input dq then dqx and dqy are calculated as dq/sqrt(2) and then used in the same way as sigma_x and sigma_y in the "Use dq Data" case. But should one provide separate values for dqx and dqy? If so, directly in 1/A as suggested by the interface?
Note also that the "Custom Slit Smear" option is not active, but there is an Slit2D object in sasmodels/resolution2d.py. Should this option be added?

**Comment by butlerpd** [comment]
I will try to look at this a bit more as we had some serious issues a while back in 4.x and I was involved in sorting through them.  For a custom pinhole we agreed that it only makes sense to give a single dq/q whether for 1D or 2D.  From the 5.0.2 interface I see that for 1D it shows `dq/q%` and an input box. For 2D, 4.x has the exact same label but also has an option on the resolution of the integration I think (with very high taking a while to compute in 2D). 5.x interface has the same choices but the label reads `<dq_low>`  in 1/A - which makes no sense. I agree that that should read just the same as 1D. I do also vaguely remember the `<dQp> <dQs>` nomenclatures but cannot remember exactly what they were.

Regarding custom slit in 2D, yes it is disabled.  I vaguely remember looking at that and deciding it was too complicated and that it didn't matter too much since kratky cameras are not generally used for collecting 2D data.

**Comment by butlerpd** [comment]
Actually, now that I am thinking more about it, I suspect that what happened is that all the work to fix resolution was done on 4.x before 5.x whas out but that those changes never got moved to 5.x? we should look into that.

**Comment by gonzalezma** [comment]
This is almost done. I did some small changes and compared to SasView 4.2.2 for all the possible options (both in 1D and 2D) and got the same results, so I'll make a pull request. But I have a couple of minor questions:
1. In Pinhole 2D we have the same interface as for 1D, with the user giving just a % dq/q and then dqx = dqy = dq = %*q. Assuming dq ~ sqrt(dqx^2+dqy^2) , it would be more consistent to do dqx = dqy = %*q/sqrt(2). But I don't think that this has any practical consequence.
2. In SasView 4, the interface shows the message "Higher accuracy is very time expensive. Use it with care ...". There is no warning message in SasView 5. Is this OK or should we add it? Perhaps as a tip tool to not encumber the interface?

**Comment by butlerpd** [comment]
If you like I can take a look (and remind myself what I did for 4.x way back when and see if I remember why? :-) so feel free to assign me along with anybody else to review the PR.

As for the warning - It think it is still important -- user might be tempted to go high because why not and then be stuck watching nothing for 10 min.  The real answer is we should have better progress bars I guess.  Not sure if tooltip would be sufficient warning? but don't really know?

**Comment by butlerpd** [comment]
OK .... so I see dqp and dqs are meant to mean dq radial and dq theta respectively. Copied from my notes on some of the corrections made in 4.x
``` python
        def _get_smear_info(self):
        """
        Get the smear info from data.
        :return: self.smear_type, self.dq_l and self.dq_r,
            respectively the type of the smear, The average <dq/q> radial(p)
            and <dq/q> theta (s)s for 2D pinhole resolution in % (slit is not
            currently supported in 2D), (dq/q)_min and (dq/q)_max for 1D pinhole
            smeared data, again in %, and dxl and/or dxw for slit smeared data
            given in 1/A and assumed constant.
        """
```

Also I note that there were tests added to check that the information in dq were not none or 0 because of possible divide by 0 error (meaning we probably had that reported :-) 
```python
        #If so check that data set has smearing info and that none are zero.
        #Otherwise no smearing can be applied using smear from data (a Gaussian
        #width of zero will cause a divide by zero error)
```
and added a bunch of logging of "Data did not contain recognizable resolution info" such as this

```python
                self.msg = "2D Data did not contain recognizable " \
                           "resolution info."
                logger.info(self.msg)
```

Also looks like we replaced confusing p and s with r and phi
```python
        smear_message_2d_x_title = "<dQ/Q>p[%]:"
        smear_message_2d_y_title = "<dQ/Q>s[%]:"
```
with
```python
        smear_message_2d_x_title = "<dQ/Q>_r[%]:"
        smear_message_2d_y_title = "<dQ/Q>_phi[%]:"
```
and 
```python
        self.smear_description_2d_x.SetToolTipString(
                                        "  dQp(parallel) in q_r direction.")
        self.smear_description_2d_y = wx.StaticText(self, wx.ID_ANY,
                            smear_message_2d_y_title, style=wx.ALIGN_LEFT)
        self.smear_description_2d_y.SetToolTipString(
                                    " dQs(perpendicular) in q_phi direction.")
        self.smear_description_pin_percent_min = wx.StaticText(self, wx.ID_ANY,
                                            smear_message_pinhole_percent_min_title,
                                            style=wx.ALIGN_LEFT)
```
with
```python
        self.smear_description_2d_x.SetToolTipString(
                                        "  dQ_r q_r in polar coordinates.")
        self.smear_description_2d_y = wx.StaticText(self, wx.ID_ANY,
                            smear_message_2d_y_title, style=wx.ALIGN_LEFT)
        self.smear_description_2d_y.SetToolTipString(
                                    " dQ_phi q_phi in polar coordinates.")
        self.smear_description_pin_percent_min = wx.StaticText(self, wx.ID_ANY,
                                            smear_message_pinhole_percent_min_title,
                                            style=wx.ALIGN_LEFT)
```

**Comment by smk78** [comment]
@butlerpd wrote:
> I will try to look at this a bit more as we had some serious issues a while back in 4.x and I was involved in sorting through them. For a custom pinhole we agreed that it only makes sense to give a single dq/q whether for 1D or 2D. From the 5.0.2 interface I see that for 1D it shows `dq/q%` and an input box. For 2D, 4.x has the exact same label but also has an option on the resolution of the integration I think (with very high taking a while to compute in 2D). 5.x interface has the same choices but the label reads `<dq_low>` in 1/A - which makes no sense. I agree that that should read just the same as 1D. I do also vaguely remember the `<dQp> <dQs>` nomenclatures but cannot remember exactly what they were.

https://github.com/SasView/sasview/issues/1236 refers.

**Comment by RichardHeenan** [comment]
Does the q resolution estimator tool use the same equations and/or notation ? See also the issue here:  
https://github.com/SasView/sasview/issues/1270

**Comment by butlerpd** [comment]
@RichardHeenan. I had forgotten about that. It is related clearly but it is using a 2D detector and approaching the pinhole problem the way I think @pkienzle is indicating we are here. I think we agree on its use in the estimator but don't think we have agreement on its use here?

**Comment by butlerpd** [comment]
This has been fixed in 5.0.3.  I think this was left open because it was forgotten in the midst of trying to understand the slit vs 2D resolution question. That is a not a urgent issue and will be captured in a new issue.  Meanwhile I am closing this as a 5.0.3 fix.
